### ISSUE TITLE
Incorrect behavior of Clocked models

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#7854]
Clocked variable models in the testsuite are contained in `Modelica.Clocked` (using MSL 4.0.0) and in `Modelica_Synchronous` (using MSL 3.2.3).

Regarding `Modelica.Clocked` we can compare the [Modelica_trunk](https://libraries.openmodelica.org/branches/master/Modelica_trunk/Modelica_trunk.html) and [Modelica_trunk_cpp](https://libraries.openmodelica.org/branches/master/Modelica_trunk_cpp/Modelica_trunk_cpp.html) reports. There are 17 models failing verification with the C runtime, of which 2 also fail in the cpp runtime. One of them is `Modelica.Clocked.Examples.Elementary.ClockSignals.LogicalSample`, which is discussed in [MSL #3858](https://github.com/modelica/ModelicaStandardLibrary/issues/3858), the other one is `Modelica.Clocked.Examples.CascadeControlledDrive.SubClocked`, which is discussed in #7852.

All the remaining failing models only fail with the C runtime, because of runtime issues that need to be investigated.

Regarding Modelica_Synchronous we can compare the [Modelica_Synchronous]() and [Modelica_Synchronous_cpp](https://libraries.openmodelica.org/branches/master/Modelica_Synchronous_cpp/Modelica_Synchronous_cpp.html) reports. There are 14 models failing the verification with the C runtime, of which only 3 also fail with the cpp runtime: 

[Modelica_Synchronous.Examples.CascadeControlledDrive.AbsoluteClocks](https://libraries.openmodelica.org/branches/master/Modelica_Synchronous/files/Modelica_Synchronous_Modelica_Synchronous.Examples.CascadeControlledDrive.AbsoluteClocks.diff.html)
[Modelica_Synchronous.Examples.CascadeControlledDrive.SubClocked](https://libraries.openmodelica.org/branches/master/Modelica_Synchronous/files/Modelica_Synchronous_Modelica_Synchronous.Examples.CascadeControlledDrive.SubClocked.diff.html)
[Modelica_Synchronous.Examples.CascadeControlledDrive.SuperSampled](https://libraries.openmodelica.org/branches/master/Modelica_Synchronous/files/Modelica_Synchronous_Modelica_Synchronous.Examples.CascadeControlledDrive.SuperSampled.diff.html).

The reference files of these three models are probably faulty for some reason, that I'll investigate with @GallLeo once he's back from holiday.

All the other models only fail in the C runtime, again due to runtime issues that need to be investigated and resolved in time for the 1.19.0 release, so that we can claim proper support of MSL 4.0.0 by then.

**Comment by AnHeuermann** [comment]
Yeah, the handling of clocked variables is in some cases simply wrong. Even very easy toy examples are not handled as I would expect them to be.

I think the problem is here: 
https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/simulation/solver/synchronous.c#L175

```C
/**
 * @brief Check when next clock needs to fire.
 *
 * If next activation time is smaller then time on next step reduce step size
 * to hit activation time of clock exactly.
 *
 * @param data            Pointer to data.
 * @param solverInfo      Solver info, containing next activation time of clocks.
 */
void checkForSynchronous(DATA *data, SOLVER_INFO* solverInfo)
{
  TRACE_PUSH
  if (listLen(data->simulationInfo->intvlTimers) > 0)
  {
    SYNC_TIMER* nextTimer = (SYNC_TIMER*)listNodeData(listFirstNode(data->simulationInfo->intvlTimers));
    double nextTimeStep = solverInfo->currentTime + solverInfo->currentStepSize;

    if ((nextTimer->activationTime <= nextTimeStep + SYNC_EPS) && (nextTimer->activationTime >= solverInfo->currentTime))
    {
      solverInfo->currentStepSize = nextTimer->activationTime - solverInfo->currentTime;
      infoStreamPrint( LOG_EVENTS_V, 0, "Adjust step-size to %.15g at time %.15g to get next timer at %.15g",
                       solverInfo->currentStepSize, solverInfo->currentTime, nextTimer->activationTime );
    }
  }
  TRACE_POP
}
```

I would need to search a bit, but had some tiny examples where this error became clear.

**Comment by AnHeuermann** [comment]
Example from @paultjevdh from #8105

```modelica
test_sample
  Real u,y1,u1;
  Clock Clk_row = Clock(1, 100) "Base row clock = 6.25MHz.";
  Clock y2;
initial equation
  u=0;
equation
  der(u)=1;
  u1 = sample(u,Clk_row);
  y1 = shiftSample(u1,4);
  y2 = shiftSample(Clk_row,4);
end test_sample;
```

**Comment by AnHeuermann** [comment]
Things we need to do to solve this issue:
- [ ] Clarify https://github.com/modelica/ModelicaSpecification/issues/3040
- [ ] Fix [fireClock()](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/simulation/solver/synchronous.c#L186-L239).
- [ ] Add test cases for each base operation relevant for Synchronous
- [ ] Add a complex test case with multiple sub-clock conversion operators.
- [ ] Resolve missing `y2` of `test_sample` in result file.

@phannebohm and myself will start this next week.

**Comment by casella** [comment]
In one week we should have some results, according to @AnHeuermann at today's devmeeting

**Comment by phannebohm** [comment]
We found this model in the MLS :confused: 
```modelica
model eventClock
  Integer nextInterval(start = 2);
  Real y1(start = 0);
  Real y2(start = 0);
equation
  when Clock(2, 100) then
    y1 = previous(y1) + 1;
  end when;
  when Clock(nextInterval , 100) then
    nextInterval = previous(nextInterval) + 1;
    y2 = previous(y2) + 1;
  end when;
end eventClock;
```

![Screenshot from 2021-11-22 17-11-30](https://user-images.githubusercontent.com/16221675/142896086-51a3cc64-3d7f-449f-ad61-b1ba166ba801.png)

High frequencies seem to be a big problem.

**Comment by casella** [comment]
Sorry, what do you mean by "MLS"? I can't find this model myself. Can you pls give the full path?

**Comment by sjoelund** [comment]
> Sorry, what do you mean by "MLS"? I can't find this model myself. Can you pls give the full path?

MLS = Modelica Language Specification. https://specification.modelica.org/master/synchronous-language-elements.html#clock-constructors

**Comment by casella** [comment]
OK, I was looking for the model name and of course I couldn't find it 😄

**Comment by casella** [comment]
There are many problems that are made visible by this example

**Comment by phannebohm** [comment]
> OK, I was looking for the model name and of course I couldn't find it smile

Oops, sorry about that. We will be looking at all the snippets from the Modelica Language Specification, so that's where things will be from. I'll put the chapter numbering in the model name.

> There are many problems that are made visible by this example

Exactly. Instead of fixing them one by one, we (mostly @AnHeuermann) will probably rewrite large parts of the code from scratch as soon as we have a proper grasp of clocks. Hence the MLS snippets :smile:

**Comment by casella** [comment]
The first problem is that clocked variables are only defined when the clock ticks. They are undefined at any other point in time, unlike discrete variables that remain constant in-between events. Dymola represents them as round dots, and only keeps a very tiny, almost invisible constant thread after them.
![screenshot](https://user-images.githubusercontent.com/4233148/143061951-1df4b6b6-dc79-46d1-ae80-c7ae5b317031.PNG)

I think we should probably do something like this.

**Comment by casella** [comment]
The second problem is that our simulations actually go backward in time, at times. If I simulate your example with 250 intervals, to make the thing even clearer, I get this:
![immagine](https://user-images.githubusercontent.com/4233148/143062249-afd6d7cf-fc01-468e-94ad-e79efec6c69a.png)
which clearly makes little sense. It seems that the output is going back by one reporting interval after each clock tick.

**Comment by casella** [comment]
The third problem is that `previous(y1)` is not really recorded correctly. If I simulate with 5000 intervals, I get this crazy stuff:
![immagine](https://user-images.githubusercontent.com/4233148/143062660-3105bd1f-b4be-466e-a194-d87e46b29b88.png)
Obviously there's something badly broken there.

Yes, Dymola does not record `previous()` values. It would be probably nice to have, because it is somewhat corresponding to der() in continuous time (i.e. gives information about dynamic behaviour), but I would strongly suggest for the time being to simply not record it. No information is (way) better than badly wrong information. Or, if you prefer: "wovon man nicht reden kann, darueber muss man schweigen" (L. Wittgenstein) 😄

**Comment by phannebohm** [comment]
> The first problem is that clocked variables are only defined when the clock ticks. They are undefined at any other point in time, unlike discrete variables that remain constant in-between events.

I'm not sure if that is a problem per se. But we may want to handle clocked variables differently in the result files, and perhaps at some point in the new runtime as well.

> Dymola represents them as round dots, and only keeps a very tiny, almost invisible constant thread after them.
> [screenshot]
> I think we should probably do something like this.

Yes! @AnHeuermann and myself talked about that.
Can we steal their plots? If not, we could plot them like they do in the MLS:
![Screenshot from 2021-11-23 17-29-02](https://user-images.githubusercontent.com/16221675/143064422-ff87e35c-590b-4a8d-90dd-466882085d1b.png)

Anyway, this requires functioning C code first...

**Comment by casella** [comment]
> > The first problem is that clocked variables are only defined when the clock ticks. They are undefined at any other point in time, unlike discrete variables that remain constant in-between events.
> 
> I'm not sure if that is a problem per se. But we may want to handle clocked variables differently in the result files, and perhaps at some point in the new runtime as well.

Sure. I'm not sure how Dymola handles them, I guess everybody will copy the same stuff, did you have a look? In the standard file there are two tables, one for constant variables and one for time-varying tables, I guess here there will also be one for each clock.

> Yes! @AnHeuermann and myself talked about that. Can we steal their plots? If not, we could plot them like they do in the MLS: ![Screenshot from 2021-11-23 17-29-02](https://user-images.githubusercontent.com/16221675/143064422-ff87e35c-590b-4a8d-90dd-466882085d1b.png)

That's pretty much what Dymola does. Let's keep it.

**Comment by casella** [comment]
@AnHeuermann, @phannebohm, after discussing this issue during today's devmeeting, we decided to keep this out of 1.19.0, because it's too late for that. Also, much better to push the fixes that are somehow correlated at the same time.

I set this tentatively for 1.19.1 with a tentative date Apr 4, because it would be nice to have this done before July 2022.  If you're done earlier, we can anticipate that release.

**Comment by AnHeuermann** [comment]
I need some clarification on the [inferred clock constructor](https://specification.modelica.org/master/synchronous-language-elements.html#clock-constructors) `Clock()`. 

Can there be different cases for a inferred clock as in 
```Modelcia
    model inferedClock
      Integer y(start=0);
    equation
      when Clock() then
        y = previous(y) + 1;
      end when;
    end inferedClock;
```
and is it correct that I can always replace `Clock()` with `Clock(resolution=1, intervalCounter=1)`?

**Comment by casella** [comment]
@AnHeuermann, this is what Dymola does, but to be honest I don't know where in the specification this behaviour is described. IMHO it is questionable.

As I understand it, `Clock()` specifies a clock that must be inferred from somewhere else. It could be the clock of some variable appearing in one of the equations, e.g. the clock of the input `u` in the example of Section 16.3, operator 16.1. If there is nowhere to infer this clock from, IMHO the compiler should report an error.

According to Section 16.3, operator 16.2, the resolution input defaults to one, but I can't see a default for the interval counter, in particular there is no default value of one for it.

**Comment by AnHeuermann** [comment]
>As I understand it, Clock() specifies a clock that must be inferred from somewhere else. It could be the clock of some variable appearing in one of the equations, e.g. the clock of the input u in the example of Section 16.3, operator 16.1. If there is nowhere to infer this clock from, IMHO the compiler should report an error.

So I guess we should also be able to handle something like this:

```modelica
model inferedClock2
  Integer tempVar(start=0);
  Integer y(start=0);
equation
  when Clock(0.1) then
     tempVar = 1;
  end when;
  when Clock() then
    y = previous(y) + tempVar;
  end when;
end inferedClock2;
```

In that case my approach to just replace `Clock()` with `Clock(1,1)` is wrong. Then the backend should replace `Clock()` with `Clock(0.1)`.
Dymola can handle that model.

**Comment by casella** [comment]
Absolutely! The second clock will be inferred to be the same as the first, and both when equations should end up in the same clocked partition, with a period of 0.1 seconds.

**Comment by casella** [comment]
Putting this back to 1.19.0, let's see how far you can go.

**Comment by AnHeuermann** [comment]
Now only `hold` and `previous` are still not working.

From the [hold operator definition](https://specification.modelica.org/master/synchronous-language-elements.html#modelica:clocked-sample)
>Before the first clock activation of u , the operator returns the start value of u, see section 16.9.
[Since the input argument is not defined before the first tick of the clock of u , the restriction is present, that it must be a component expression (or a parameter expression), in order that the initial value of  u can be used in such a case.]

I understand that for the following model

```modelica
model holdVar
  Clock clk1 = Clock(0.3);  // ticks: 0, 0.3, 0.6, 0.9, ...
  Real xc(start = 0), x;
equation 
  when clk1 then
    xc = previous(xc) + 1;
  end when;
  x = hold(xc);
end holdVar;
```

`x` needs to be 0.0 at initialization and `1.0` at the first tick at simulation time `0.0`. 
Is this correct?

Dymola will only save the value `1.0` at time `0.0`:
![image](https://user-images.githubusercontent.com/38031952/145984786-1ed2de25-e6bd-4cb0-870d-8670b6efe8a4.png)

**Comment by AnHeuermann** [comment]
So I think I'm now at a point where it won't get much better unless we start to save "normal" variables and clocked variables differently. When a variable isn't defined yet (because the clock didn't tick for the first time) it doesn't make sense to save it to the result file.

I'll open a PR and see how much I'll break.

**Comment by sjoelund** [comment]
C++ runtime is also running at half the speed compared to the C runtime. It's using templates quite frequently.

**Comment by sjoelund** [comment]
> @sjoelund, at least for the time being, that would be a good reason to use the cpp runtime instead of the default c runtime.

I tested a library we have worked a lot on lately... Buildings has 1257 models simulating with C runtime but only 978 with C++ runtime. So it's a bit misleading to see the coverage of the libraries that C++ runtime developers worked on compared to the ones the C runtime developers worked on. So it takes twice as long to compile models and which libraries work with it is a bit spotty. That makes it a worse candidate for default runtime. Does it support source-code FMUs? That's another convenient feature for generating Windows-compatible FMUs from Linux.

**Comment by casella** [comment]
I'm not advocating to switch to the cpp runtime by default. What I meant was that, unless @AnHeuermann manages to fix most of the Clock issues in the C runtime (which is the topic of this ticket), I guess we should mention in the release notes that the cpp runtime should preferably be used for proper synchronous system support in 1.19.0. That's a fact.

@sjoelund do we run tests with cpp on a regular basis over the whole library testsuite? I guess we should have such a test, e.g. master-cpp

**Comment by sjoelund** [comment]
Nope, we only have a few libraries tested with C++. You need different flags for everything, so you cannot easily just test all the libraries with C++.
If the interface was the same as the C runtime it would be a lot easier, but student flags, result files, compiler flags, etc.

**Comment by casella** [comment]
> Nope, we only have a few libraries tested with C++. You need different flags for everything, so you cannot easily just test all the libraries with C++. If the interface was the same as the C runtime it would be a lot easier, but student flags, result files, compiler flags, etc.

Let's discuss it in #8330, this is off-topic here as long as synchronous systems are not specifically involved.

**Comment by AnHeuermann** [comment]
#8324 is now merged to the master 🎉 
Let's wait to see if the nightly tests show improvements for the C runtime and no regressions for the C++ runtime.

**Comment by casella** [comment]
Excellent! 😃

**Comment by sjoelund** [comment]
Seems good:
https://libraries.openmodelica.org/branches/history/master/2022-01-02%2022:31:55..2022-01-04%2022:23:11.html
https://libraries.openmodelica.org/branches/history/cpp/2022-01-02%2020:54:12..2022-01-04%2020:54:34.html

**Comment by phannebohm** [comment]
Regarding `Modelica.Clocked` apart from `Modelica.Clocked.Examples.Elementary.ClockSignals.LogicalSample` everything seems to be fine both in C and Cpp.

Regarding `Modelica_Synchronous` both runtimes are now only failing verification for the same 3 models mentioned at the top of this issue. Would be nice to know if those are actually wrong or if it's the reference files.

There are still a few things left to do, in particular #4010 #4363 #5404 and #8349.

**Comment by casella** [comment]
> Regarding `Modelica.Clocked` apart from `Modelica.Clocked.Examples.Elementary.ClockSignals.LogicalSample` everything seems to be fine both in C and Cpp.

The issue with LogicalSample is only on the [cpp runtime](https://libraries.openmodelica.org/branches/cpp/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Clocked.Examples.Elementary.ClockSignals.LogicalSample.diff.html). @rfranke, maybe you can check how this is handled in the C runtime now.

> Regarding `Modelica_Synchronous` both runtimes are now only failing verification for the same 3 models mentioned at the top of this issue. Would be nice to know if those are actually wrong or if it's the reference files.

`Modelica_Synchronous` was last released almost two years ago and is [deprecated](https://github.com/modelica/Modelica_Synchronous) since the release of MSL 4.0.0. The three failing examples have been ported to `Modelica.Clocked.Examples.CascadeControlledDrive`, which pass verification with flying colours using both runtimes.

My recommendation is not to waste time on this, and just remove `Modelica_Synchronous` from the testsuite and package manager. There is no reason why we should still use it for any purpose. See #8380 and https://github.com/OpenModelica/OMPackageManager/pull/7.

**Comment by casella** [comment]
> There are still a few things left to do, in particular #4010

This is an improvement, and is probably somehow a duplicate of #2152.

> #4363 

This is definitely obsolete, I just closed it with some comments.

>#5404 

This is kind of weird, we should definitely look into that.

> and #8349.

This should also be fixed.

Summing up, the situation with clocked models has improved dramatically with the latest commits. We still have two pending issues regarding correctness, namely #5404 and #8349. If those are quick fixes, it would be nice to get them into 1.19.0. Otherwise we can close this ticket and take care of them in 1.20.0

What do you think?

**Comment by phannebohm** [comment]
Okay great! Yes, currently working on those two issues. #8349 is not much work I think, no idea on #5404 yet.
Anyway I agree, there is no need to wait, so I'll just close this one.
